Nsight Perf SDK performance counters for D3D12, D3D11, Vulkan, OpenGL, Windows, and Linux

jglines-nvidia opened the pull request:
This implements performance counters using the NVIDIA Nsight Perf SDK, supporting the latest NVIDIA GPUs and D3D12, D3D11, Vulkan, OpenGL, Windows, and Linux. With this patch, when a user with an NVIDIA GPU first opens the performance counter selection dialog in RenderDoc they will find the following message directing them to download and install the Nsight Perf SDK library. After installing the library and reopening the capture, Nsight Perf SDK counters become available.

![renderdoc_nvperf_missing_library](https://user-images.githubusercontent.com/68713353/208970535-fe7f4e03-330d-41a4-b9ee-20e221663d59.png)

![renderdoc_nvperf_counters](https://user-images.githubusercontent.com/68713353/208970624-4e2be607-ed34-49b7-9647-ebdd85f06419.png)

jglines-nvidia committed:
Collect Nsight Perf SDK counters in D3D11

This adds the NVD3D11Counters class, which implements D3D11 counter
collection on NVIDIA hardware via the NVIDIA Nsight Perf SDK.

jglines-nvidia committed:
Collect Nsight Perf SDK counters in Vulkan

This adds the NVVulkanCounters class, which implements Vulkan counter
collection on NVIDIA hardware via the NVIDIA Nsight Perf SDK.

Some modifications are made to the NvPerfUtility headers in order to
use the Vulkan dispatch tables from RenderDoc.

On Windows, the VK_KHR_external_semaphore_win32 extension is required
for synchronization with the NvPerf "service pending GPU operations"
thread.

jglines-nvidia committed:
Collect Nsight Perf SDK counters in OpenGL

This adds the NVGLCounters class, which implements OpenGL counter
collection on NVIDIA hardware via the NVIDIA Nsight Perf SDK.

jglines-nvidia committed:
Enable external links in counter selection UI

Use QTextBrowser for the counter description widget, which allows for opening
links in an external browser.

This change is needed to make the link to NVIDIA Nsight Perf SDK clickable.

jglines-nvidia committed:
Collect Nsight Perf SDK counters in D3D12

This adds the NVD3D12Counters class, which implements D3D12 counter
collection on NVIDIA hardware via the NVIDIA Nsight Perf SDK.

A helper class NVCounterEnumerator is added to provide functionality
common to NVIDIA counters for other graphics APIs.

jglines-nvidia committed:
Fix performance counters without a category

In the performance counter selection dialog, show counters without a category
at root of tree, rather than as children of a nameless "" category.

jglines-nvidia committed:
Add MetricsConfigBuilder::GetNumPasses() to NvPerfUtility

This method is used to calculate the expected maximum number of replays and
enforce an upper-bound on the replay loop.

jglines-nvidia committed:
Load dxgi function at runtime in NvPerfUtility

This is to avoid linking with dxgi at build time.

jglines-nvidia committed:
Avoid gcc unused argument warning in NvPerfUtility

jglines-nvidia committed:
Hide Vulkan symbols in NvPerfUtility

Adds the macro NV_PERF_UTILITY_HIDE_VULKAN_SYMBOLS to NvPerfUtility, which
allows for passing Vulkan function pointers without linking with Vulkan symbols
that are not present in RenderDoc.

baldurk commented:
This will need to be rebased onto latest now that the precursor PR is merged. FYI for the future it's best to PR dependent changes one at a time.

jglines-nvidia committed:
Fix VS2015 compile errors for NvPerfUtility headers

All of these compile errors are from implicit bool conversions.

jglines-nvidia committed:
Fix Clang 12 compile errors for NvPerfUtility headers

Compile errors include missing EOL newline, missing override keyword for member
function, and implicit long to int conversion.

jglines-nvidia commented:
This is rebased now, and all of the CI is passing except for what looks like some unrelated failure installing Mac Homebrew dependencies.
